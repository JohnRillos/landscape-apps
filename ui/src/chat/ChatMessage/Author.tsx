import React from 'react';
import {
  makePrettyDayAndDateAndTime,
  makePrettyDayAndTime,
  makePrettyTime,
} from '@/logic/utils';
import { useLocation } from 'react-router';
import { useModalNavigate } from '@/logic/routing';
import Avatar from '@/components/Avatar';
import ShipName from '@/components/ShipName';

interface AuthorProps {
  ship: string;
  date?: Date;
  timeOnly?: boolean;
  hideTime?: boolean;
}
export default function Author({
  ship,
  date,
  timeOnly,
  hideTime,
}: AuthorProps) {
  const location = useLocation();
  const modalNavigate = useModalNavigate();
  const prettyTime = date ? makePrettyTime(date) : undefined;
  const prettyDayAndTime = date ? makePrettyDayAndTime(date) : undefined;
  const prettyDayAndDateAndTime = date
    ? makePrettyDayAndDateAndTime(date)
    : undefined;

  const handleProfileClick = () => {
    modalNavigate(`/profile/${ship}`, {
      state: { backgroundLocation: location },
    });
  };

  if (!date) {
    return (
      <div className="align-center group flex items-center space-x-3 py-1">
        <div onClick={handleProfileClick}>
          <Avatar ship={ship} size="xs" className="cursor-pointer" />
        </div>
        <ShipName name={ship} showAlias className="text-md font-semibold" />
      </div>
    );
  }

  return (
    <div className="align-center group flex items-center space-x-3 py-1">
      <div onClick={handleProfileClick}>
        <Avatar ship={ship} size="xs" className="cursor-pointer" />
      </div>
      <ShipName name={ship} showAlias className="text-md font-semibold" />
      {hideTime ? (
        <span className="hidden text-sm font-semibold text-gray-500 group-hover:block">
          {prettyDayAndTime}
        </span>
      ) : (
        <>
          <span className="hidden text-sm font-semibold text-gray-500 group-hover:block">
            {prettyDayAndDateAndTime}
          </span>
          <span className="block text-sm font-semibold text-gray-500 group-hover:hidden">
            {timeOnly ? prettyTime : prettyDayAndTime}
          </span>
        </>
      )}
    </div>
  );
}
