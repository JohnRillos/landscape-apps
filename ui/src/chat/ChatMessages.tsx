import React, { useEffect } from 'react';
import { BigIntOrderedMap } from '@urbit/api';

import { ChatMessageProps } from './ChatMessage/ChatMessage';
import { ChatWrit } from '../types/chat';
import VirtuosoScroller from './ChatScroller/VirtuosoScroller';
import Groups1Scroller from './ChatScroller/Groups1Scroller';

interface ChatMessagesProps
  extends Omit<
    ChatMessageProps,
    'writ' | 'newAuthor' | 'newDay' | 'time' | 'whom'
  > {
  whom: string;

  messages: BigIntOrderedMap<ChatWrit>;
  replying?: boolean;
}

export default function ChatMessages(props: ChatMessagesProps) {
  const { messages, replying = false, whom } = props;

  const Scroller =
    window.scroller === 'virtuoso' ? VirtuosoScroller : Groups1Scroller;

  return <Scroller {...{ whom, messages, replying }} />;
}
