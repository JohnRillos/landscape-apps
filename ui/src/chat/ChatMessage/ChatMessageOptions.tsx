import React, { useCallback } from 'react';
import { useNavigate } from 'react-router';
import { useChatState } from '../../state/chat';
import { ChatWrit } from '../../types/chat';
import IconButton from '../../components/IconButton';
import BubbleIcon from '../../components/icons/BubbleIcon';
import EllipsisIcon from '../../components/icons/EllipsisIcon';
import FaceIcon from '../../components/icons/FaceIcon';
import HashIcon from '../../components/icons/HashIcon';
import ShareIcon from '../../components/icons/ShareIcon';
import XIcon from '../../components/icons/XIcon';
import { useChatStore } from '../useChatStore';

export default function ChatMessageOptions(props: {
  whom: string;
  writ: ChatWrit;
}) {
  const { whom, writ } = props;
  const onDelete = () => {
    useChatState.getState().delMessage(whom, writ.seal.id);
  };

  const reply = useCallback(() => {
    useChatStore.getState().reply(whom, writ.seal.id);
  }, [writ, whom]);

  const navigate = useNavigate();
  return (
    <div className="absolute right-2 -top-5 z-10 flex flex space-x-0.5 rounded-lg border border-gray-100 bg-white p-[1px] align-middle opacity-0 group-one-hover:opacity-100">
      <IconButton
        icon={<FaceIcon className="h-6 w-6 text-gray-400" />}
        label="React"
        showTooltip
        action={() => console.log('react')}
      />
      {!writ.memo.replying ? (
        <>
          <IconButton
            icon={<BubbleIcon className="h-6 w-6 text-gray-400" />}
            label="Reply"
            showTooltip
            action={reply}
          />
          <IconButton
            icon={<HashIcon className="h-6 w-6 text-gray-400" />}
            label="Start Thread"
            showTooltip
            action={() => navigate(`message/${writ.seal.id}`)}
          />
        </>
      ) : null}
      <IconButton
        icon={<ShareIcon className="h-6 w-6 text-gray-400" />}
        label="Send to..."
        showTooltip
        action={() => console.log('send to..')}
      />
      {window.our === writ.memo.author ? (
        <IconButton
          icon={<XIcon className="h-6 w-6 text-red" />}
          label="Delete"
          showTooltip
          action={onDelete}
        />
      ) : null}

      <IconButton
        icon={<EllipsisIcon className="h-6 w-6 text-gray-400" />}
        label="More..."
        showTooltip
        action={() => console.log('More...')}
      />
    </div>
  );
}
