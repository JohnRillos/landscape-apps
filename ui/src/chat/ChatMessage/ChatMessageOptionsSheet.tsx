import cn from 'classnames';
import ConfirmationModal from '@/components/ConfirmationModal';
import BubbleIcon from '@/components/icons/BubbleIcon';
import CheckIcon from '@/components/icons/CheckIcon';
import CopyIcon from '@/components/icons/CopyIcon';
import FaceIcon from '@/components/icons/FaceIcon';
import HashIcon from '@/components/icons/HashIcon';
import XIcon from '@/components/icons/XIcon';
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner';
import Sheet, { SheetContent } from '@/components/Sheet';
import SidebarItem from '@/components/Sidebar/SidebarItem';
import useRequestState from '@/logic/useRequestState';
import { canWriteChannel, useCopy } from '@/logic/utils';
import { useChatPerms, useChatState } from '@/state/chat';
import useEmoji from '@/state/emoji';
import { useAmAdmin, useGroup, useRouteGroup, useVessel } from '@/state/groups';
import { ChatWrit } from '@/types/chat';
import Picker from '@emoji-mart/react';
import { useCallback, useEffect, useState } from 'react';
import { useParams, useSearchParams } from 'react-router-dom';

export default function ChatMessageOptionsSheet({
  sheetOpen,
  setSheetOpen,
  whom,
  writ,
  hideReply = false,
  hideThreadReply = false,
}: {
  sheetOpen: boolean;
  setSheetOpen: (value: boolean) => void;
  whom: string;
  writ: ChatWrit;
  hideReply?: boolean;
  hideThreadReply?: boolean;
}) {
  const { data, load } = useEmoji();
  const [showPicker, setShowPicker] = useState(false);
  const [deleteOpen, setDeleteOpen] = useState(false);
  const width = window.innerWidth;
  const mobilePerLineCount = Math.floor((width - 20) / 36);
  const [_searchParams, setSearchParams] = useSearchParams();
  const {
    isPending: isDeletePending,
    setPending: setDeletePending,
    setReady,
  } = useRequestState();
  const { didCopy, doCopy } = useCopy(
    `/1/chan/chat/${whom}/msg/${writ.seal.id}`
  );
  const { chShip, chName } = useParams();
  const chFlag = `${chShip}/${chName}`;
  const groupFlag = useRouteGroup();
  const group = useGroup(groupFlag);
  const perms = useChatPerms(chFlag);
  const vessel = useVessel(groupFlag, window.our);
  const canWrite = canWriteChannel(perms, vessel, group?.bloc);
  const isAdmin = useAmAdmin(groupFlag);

  useEffect(() => {
    load();
  }, [load]);

  const onDelete = async () => {
    setDeletePending();
    try {
      await useChatState.getState().delMessage(whom, writ.seal.id);
    } catch (e) {
      console.log('Failed to delete message', e);
    }
    setReady();
  };

  const onCopy = useCallback(() => {
    doCopy();
  }, [doCopy]);

  const reply = useCallback(() => {
    setSearchParams({ chat_reply: writ.seal.id }, { replace: true });
  }, [writ, setSearchParams]);

  const onEmoji = useCallback(
    (emoji: { shortcodes: string }) => {
      useChatState.getState().addFeel(whom, writ.seal.id, emoji.shortcodes);
      setSheetOpen(false);
    },
    [whom, writ, setSheetOpen]
  );

  const actionWrapper = useCallback(
    (action: () => void) => () => {
      action();
      setSheetOpen(false);
    },
    [setSheetOpen]
  );

  return (
    <Sheet open={sheetOpen} onOpenChange={(o) => setSheetOpen(o)}>
      <SheetContent className={cn({ hidden: deleteOpen })}>
        {!showPicker ? (
          <div className="flex flex-col pt-4">
            {canWrite ? (
              <SidebarItem
                icon={
                  <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-50">
                    <FaceIcon className="h-6 w-6" />
                  </div>
                }
                onClick={() => setShowPicker(true)}
              >
                Add Reaction
              </SidebarItem>
            ) : null}
            {!hideReply ? (
              <SidebarItem
                icon={
                  <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-50">
                    <BubbleIcon className="h-6 w-6" />
                  </div>
                }
                onClick={actionWrapper(reply)}
              >
                Reply
              </SidebarItem>
            ) : null}
            {!writ.memo.replying &&
            writ.memo.replying?.length !== 0 &&
            !hideThreadReply ? (
              <SidebarItem
                icon={
                  <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-50">
                    <HashIcon className="h-6 w-6" />
                  </div>
                }
                to={`message/${writ.seal.id}`}
              >
                Thread Reply
              </SidebarItem>
            ) : null}
            {groupFlag ? (
              <SidebarItem
                icon={
                  <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-50">
                    {didCopy ? (
                      <CheckIcon className="h-6 w-6" />
                    ) : (
                      <CopyIcon className="h-6 w-6" />
                    )}
                  </div>
                }
                onClick={actionWrapper(onCopy)}
              >
                Copy
              </SidebarItem>
            ) : null}
            {isAdmin || window.our === writ.memo.author ? (
              <SidebarItem
                icon={
                  <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-50">
                    <XIcon className="h-6 w-6 text-red" />
                  </div>
                }
                onClick={() => setDeleteOpen(true)}
              >
                Delete
              </SidebarItem>
            ) : null}
          </div>
        ) : (
          <div className="flex h-[200px] w-full flex-col items-center justify-center space-y-2 pt-8">
            {data ? (
              <Picker
                data={data}
                perLine={mobilePerLineCount}
                previewPosition="none"
                navPosition="bottom"
                searchPosition="none"
                onEmojiSelect={onEmoji}
                noCountryFlags={false}
              />
            ) : (
              <div className="flex h-full w-full items-center justify-center">
                <LoadingSpinner className="h-6 w-6" />
              </div>
            )}
          </div>
        )}
        <ConfirmationModal
          title="Delete Message"
          message="Are you sure you want to delete this message?"
          onConfirm={onDelete}
          open={deleteOpen}
          setOpen={setDeleteOpen}
          confirmText="Delete"
          loading={isDeletePending}
        />
      </SheetContent>
    </Sheet>
  );
}
