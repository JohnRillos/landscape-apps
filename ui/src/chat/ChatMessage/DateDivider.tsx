import React from 'react';
import cn from 'classnames';
import { makePrettyDay, pluralize } from '@/logic/utils';
import { useChatState } from '@/state/chat';

interface DateDividerProps {
  date: Date;
  unreadCount?: number;
  viewRef?: (node?: Element | null | undefined) => void;
}

export default function DateDivider({
  date,
  unreadCount,
  viewRef,
}: DateDividerProps) {
  const prettyDay = makePrettyDay(date);

  return (
    <div ref={viewRef} className="flex w-full items-center py-4">
      <div
        className={cn(
          'h-[2px] w-8 rounded-sm',
          unreadCount ? 'bg-blue' : 'bg-gray-200'
        )}
      >
        &nbsp;
      </div>
      <span
        className={cn(
          'whitespace-nowrap px-3 font-semibold',
          unreadCount ? 'text-blue' : 'text-gray-400'
        )}
      >
        {prettyDay}
        {unreadCount ? (
          <>
            &nbsp;&bull;&nbsp;
            {unreadCount} new {pluralize('messsage', unreadCount)} below
          </>
        ) : null}
      </span>
      <div
        className={cn(
          'h-[2px] w-full rounded-sm',
          unreadCount ? 'bg-blue' : 'bg-gray-200'
        )}
      >
        &nbsp;
      </div>
    </div>
  );
}
