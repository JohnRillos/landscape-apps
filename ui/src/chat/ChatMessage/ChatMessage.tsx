import React from 'react';
import cn from 'classnames';
import _ from 'lodash';
import f from 'lodash/fp';
import { BigInteger } from 'big-integer';
import { daToUnix } from '@urbit/api';
import { format, formatDistanceToNow, formatRelative, isToday } from 'date-fns';
import { NavLink } from 'react-router-dom';
import { ChatBrief, ChatWrit } from '../../types/chat';
import Author from './Author';
import ChatContent from '../ChatContent/ChatContent';
import ChatReactions from '../ChatReactions/ChatReactions';
import DateDivider from './DateDivider';
import ChatMessageOptions from './ChatMessageOptions';
import { usePact } from '../../state/chat';
import Avatar from '../../components/Avatar';

export interface ChatMessageProps {
  whom: string;
  time: BigInteger;
  writ: ChatWrit;
  isReplyOp?: boolean;
  newAuthor?: boolean;
  newDay?: boolean;
  unread?: ChatBrief;
  hideReplies?: boolean;
}

const ChatMessage = React.forwardRef<HTMLDivElement, ChatMessageProps>(
  (
    {
      whom,
      time,
      writ,
      unread,
      isReplyOp = false,
      newAuthor = false,
      newDay = false,
      hideReplies = false,
    }: ChatMessageProps,
    ref
  ) => {
    const { seal, memo } = writ;

    const unix = new Date(daToUnix(time));

    const pact = usePact(whom);

    const numReplies = seal.replied.length;
    const replyAuthors = _.flow(
      f.map((k: string) => {
        const t = pact.index[k];
        const mess = t ? pact.writs.get(t) : undefined;
        if (!mess) {
          return undefined;
        }
        return mess.memo.author;
      }),
      f.compact,
      f.uniq,
      f.take(3)
    )(seal.replied);

    return (
      <div ref={ref} className="flex flex-col">
        {unread ? <DateDivider date={unix} unreadCount={unread.count} /> : null}
        {newDay && !unread ? <DateDivider date={unix} /> : null}
        {newAuthor ? <Author ship={memo.author} date={unix} /> : null}
        <div className="group-one relative z-0 flex">
          <ChatMessageOptions whom={whom} writ={writ} />
          <div className="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 group-one-hover:opacity-100">
            {format(unix, 'HH:mm')}
          </div>
          <div
            className={cn(
              'flex w-full flex-col space-y-2 rounded py-1 pl-3 pr-2 group-one-hover:bg-gray-50',
              isReplyOp && 'bg-gray-50'
            )}
          >
            {'story' in memo.content ? (
              <ChatContent story={memo.content.story} />
            ) : null}
            {Object.keys(seal.feels).length > 0 && (
              <ChatReactions seal={seal} />
            )}
            {numReplies > 0 && !hideReplies ? (
              <NavLink
                to={`message/${seal.id}`}
                className={({ isActive }) =>
                  cn(
                    'rounded p-2 text-sm font-semibold text-blue',
                    isActive ? 'bg-gray-50' : ''
                  )
                }
              >
                <div className="flex items-center space-x-2">
                  {replyAuthors.map((ship) => (
                    <Avatar key={ship} ship={ship} size="xs" />
                  ))}

                  <span>
                    {numReplies} {numReplies > 1 ? 'replies' : 'reply'}{' '}
                  </span>
                  <span className="text-gray-400">
                    Last reply{' '}
                    {isToday(unix)
                      ? `${formatDistanceToNow(unix)} ago`
                      : formatRelative(unix, new Date())}
                  </span>
                </div>
              </NavLink>
            ) : null}
          </div>
        </div>
      </div>
    );
  }
);

export default ChatMessage;
