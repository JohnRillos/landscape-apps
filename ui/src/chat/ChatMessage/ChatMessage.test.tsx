import { describe, it, expect, beforeEach } from 'vitest';
import React from 'react';
import { MemoryRouter } from 'react-router';
import { render } from '@testing-library/react';
import { unixToDa } from '@urbit/api';
import ChatMessage from './ChatMessage';
import { makeFakeChatWrit } from '../../mocks/chat';

describe('ChatMessage', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });
  it('renders as expected', () => {
    const date = new Date(2021, 1, 1, 13);
    const writ = makeFakeChatWrit(
      1,
      '~finned-palmer',
      {
        block: [],
        inline: [{ bold: ['A bold test message'] }, 'with some more text'],
      },
      undefined
    );
    const da = unixToDa(date.valueOf());
    const { asFragment } = render(
      <MemoryRouter>
        <ChatMessage time={da} whom="~zod/test" writ={writ} newAuthor newDay />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
