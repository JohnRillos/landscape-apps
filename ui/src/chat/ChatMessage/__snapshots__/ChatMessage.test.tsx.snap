// Vitest Snapshot v1

exports[`ChatMessage > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <div
      class="flex w-full items-center py-4"
    >
      <div
        class="h-[2px] w-8 rounded-sm bg-gray-200"
      >
         
      </div>
      <span
        class="whitespace-nowrap px-3 font-semibold text-gray-400"
      >
        February 1st
      </span>
      <div
        class="h-[2px] w-full rounded-sm bg-gray-200"
      >
         
      </div>
    </div>
    <div
      class="align-center group flex items-center space-x-3 py-1"
    >
      <div
        class="relative flex flex-none items-center justify-center rounded bg-black w-6 h-6 rounded p-1.5"
        style="background-color: rgb(0, 0, 0);"
      >
        <svg
          class=""
          style="display: block; width: 12px; height: 12px;"
          version="1.1"
          viewBox="0 0 12 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="#000000"
            height="12px"
            width="12px"
            x="0px"
            y="0px"
          />
          <g
            transform="matrix(0.046875,0,0,0.046875,0,0)"
          >
            <g
              fill="none"
              transform="matrix(1,0,0,1,0,0)"
            >
              <rect
                dataisgeon="true"
                fill="white"
                height="128"
                stroke="#000000"
                stroke-linecap="square"
                stroke-width="0.8px"
                vector-effect="non-scaling-stroke"
                width="128"
              />
            </g>
            <g
              fill="none"
              transform="matrix(1,0,0,1,128,0)"
            >
              <circle
                cx="64"
                cy="64"
                dataisgeon="true"
                fill="white"
                r="64"
                stroke="#000000"
                stroke-linecap="square"
                stroke-width="0.8px"
                vector-effect="non-scaling-stroke"
              />
            </g>
            <g
              fill="none"
              transform="matrix(1,0,0,1,0,128)"
            >
              <path
                d="M0.0541 0C70.7217 0.0292317 128 57.3256 128 128C57.3177 128 0.0164917 70.7089 7.62806e-06 0.0305091C7.62851e-06 0.0203397 -4.44317e-10 0.01017 0 0H0.0541Z"
                dataisgeon="true"
                fill="white"
                stroke="#000000"
                stroke-linecap="square"
                stroke-width="0.8px"
                vector-effect="non-scaling-stroke"
              />
            </g>
            <g
              fill="none"
              transform="matrix(1,0,0,1,128,128)"
            >
              <path
                d="M64 128H0L5.59506e-06 0L64 5.59506e-06C99.3462 8.68512e-06 128 28.6538 128 64C128 99.3462 99.3462 128 64 128Z"
                dataisgeon="true"
                fill="white"
                stroke="#000000"
                stroke-linecap="square"
                stroke-width="0.8px"
                vector-effect="non-scaling-stroke"
              />
            </g>
          </g>
        </svg>
      </div>
      <span
        class="text-md font-semibold"
      >
        <span
          aria-hidden="true"
        >
          ~
        </span>
        <span>
          finned
        </span>
        <span
          aria-hidden="true"
        >
          -
        </span>
        <span
          aria-hidden="false"
        >
          palmer
        </span>
      </span>
      <span
        class="hidden text-sm font-semibold text-gray-500 group-hover:block"
      >
        February 1st, 2021 • 13:00
      </span>
      <span
        class="block text-sm font-semibold text-gray-500 group-hover:hidden"
      >
        February 1st • 13:00
      </span>
    </div>
    <div
      class="group-one relative z-0 flex"
    >
      <div
        class="absolute right-2 -top-5 z-10 flex flex space-x-0.5 rounded-lg border border-gray-100 bg-white p-[1px] align-middle opacity-0 group-one-hover:opacity-100"
      >
        <div
          class="group-two relative cursor-pointer"
        >
          <div
            class="z-2 pointer-events-none absolute -top-10 grid grid-cols-1 grid-rows-2 justify-items-center rounded opacity-0 group-two-hover:opacity-100"
          >
            <div
              class="hidden w-fit cursor-none rounded bg-gray-400 px-4 py-2 group-two-hover:block"
            >
              <label
                class="whitespace-nowrap font-semibold text-white"
              >
                React
              </label>
            </div>
            <div>
              <svg
                fill="none"
                height="8"
                viewBox="0 0 17 8"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 0L0.5 0L7.08579 6.58579C7.86684 7.36684 9.13316 7.36684 9.91421 6.58579L16.5 0Z"
                  fill="#999999"
                />
              </svg>
            </div>
          </div>
          <button
            aria-label="React"
            class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50"
          >
            <svg
              class="h-6 w-6 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M7 4h10c1.6569 0 3 1.34315 3 3v10c0 1.6569-1.3431 3-3 3H7c-1.65685 0-3-1.3431-3-3V7c0-1.65685 1.34315-3 3-3Zm10-2H7C4.23858 2 2 4.23858 2 7v10c0 2.7614 2.23858 5 5 5h10c2.7614 0 5-2.2386 5-5V7c0-2.76142-2.2386-5-5-5ZM8.5 8C7.67157 8 7 8.67157 7 9.5v1c0 .8284.67157 1.5 1.5 1.5s1.5-.6716 1.5-1.5v-1C10 8.67157 9.32843 8 8.5 8ZM14 9.5c0-.82843.6716-1.5 1.5-1.5s1.5.67157 1.5 1.5v1c0 .8284-.6716 1.5-1.5 1.5s-1.5-.6716-1.5-1.5v-1ZM13 15c0 .5523-.4477 1-1 1s-1-.4477-1-1 .4477-1 1-1 1 .4477 1 1Zm2 0c0 1.6569-1.3431 3-3 3s-3-1.3431-3-3 1.3431-3 3-3 3 1.3431 3 3Z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
        <div
          class="group-two relative cursor-pointer"
        >
          <div
            class="z-2 pointer-events-none absolute -top-10 grid grid-cols-1 grid-rows-2 justify-items-center rounded opacity-0 group-two-hover:opacity-100"
          >
            <div
              class="hidden w-fit cursor-none rounded bg-gray-400 px-4 py-2 group-two-hover:block"
            >
              <label
                class="whitespace-nowrap font-semibold text-white"
              >
                Reply
              </label>
            </div>
            <div>
              <svg
                fill="none"
                height="8"
                viewBox="0 0 17 8"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 0L0.5 0L7.08579 6.58579C7.86684 7.36684 9.13316 7.36684 9.91421 6.58579L16.5 0Z"
                  fill="#999999"
                />
              </svg>
            </div>
          </div>
          <button
            aria-label="Reply"
            class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50"
          >
            <svg
              class="h-6 w-6 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                d="m4.44444 14.636-.94419.3294.94419-.3294Zm14.68296 1.001-.9806.1961-.0683-.3414.1585-.31.8904.4553Zm.7255 3.6276.9806-.1962-.9806.1962Zm-.5883.5883-.1962.9806.1962-.9806Zm-3.6276-.7255-.4553-.8904.31-.1585.3414.0683-.1961.9806Zm-6.27299.4282.32938-.9442-.32938.9442ZM5 12c0 .8099.13714 1.5857.38864 2.3066l-1.88839.6588C3.17587 14.0355 3 13.0372 3 12h2Zm7-7c-3.86599 0-7 3.13401-7 7H3c0-4.97056 4.02944-9 9-9v2Zm7 7c0-3.86599-3.134-7-7-7V3c4.9706 0 9 4.02944 9 9h-2Zm-.763 3.1817c.4875-.9535.763-2.034.763-3.1817h2c0 1.4716-.3539 2.8634-.9822 4.0922l-1.7808-.9105Zm1.871.2591.7255 3.6276-1.9612.3923-.7255-3.6276 1.9612-.3923Zm.7255 3.6276c.2099 1.0496-.7155 1.975-1.7651 1.7651l.3923-1.9612c-.3499-.0699-.6583.2385-.5884.5884l1.9612-.3923Zm-1.7651 1.7651-3.6276-.7255.3923-1.9612 3.6276.7255-.3923 1.9612ZM12 19c1.1477 0 2.2282-.2755 3.1817-.763l.9105 1.7808C14.8634 20.6461 13.4716 21 12 21v-2Zm-2.30661-.3886C10.4143 18.8629 11.1901 19 12 19v2c-1.0372 0-2.03554-.1759-2.96538-.5002l.65877-1.8884Zm-4.30475-4.3048c.70113 2.0098 2.29492 3.6036 4.30475 4.3048l-.65877 1.8884c-2.58577-.9021-4.63233-2.9487-5.53437-5.5344l1.88839-.6588Z"
              />
            </svg>
          </button>
        </div>
        <div
          class="group-two relative cursor-pointer"
        >
          <div
            class="z-2 pointer-events-none absolute -top-10 grid grid-cols-1 grid-rows-2 justify-items-center rounded opacity-0 group-two-hover:opacity-100"
          >
            <div
              class="hidden w-fit cursor-none rounded bg-gray-400 px-4 py-2 group-two-hover:block"
            >
              <label
                class="whitespace-nowrap font-semibold text-white"
              >
                Start Thread
              </label>
            </div>
            <div>
              <svg
                fill="none"
                height="8"
                viewBox="0 0 17 8"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 0L0.5 0L7.08579 6.58579C7.86684 7.36684 9.13316 7.36684 9.91421 6.58579L16.5 0Z"
                  fill="#999999"
                />
              </svg>
            </div>
          </div>
          <button
            aria-label="Start Thread"
            class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50"
          >
            <svg
              class="h-6 w-6 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M10.124 3.00771c.5481.0685.9368.56829.8683 1.11631l-.4845 3.87596h3.9844l.5155-4.12403c.0685-.54802.5683-.93675 1.1163-.86824.5481.0685.9368.56829.8683 1.11631l-.4845 3.87596H19c.5523 0 1 .44772 1 1 0 .55229-.4477 1-1 1h-2.7422L15.7578 14H19c.5523 0 1 .4477 1 1s-.4477 1-1 1h-3.4922l-.5155 4.124c-.0685.548-.5683.9368-1.1163.8683-.5481-.0685-.9368-.5683-.8683-1.1164L13.4922 16H9.50778l-.5155 4.124c-.0685.548-.56829.9368-1.11631.8683s-.93675-.5683-.86825-1.1164L7.49222 16H5c-.55228 0-1-.4477-1-1s.44772-1 1-1h2.74222l.5-4.00002H5c-.55228 0-1-.44771-1-1 0-.55228.44772-1 1-1h3.49222l.5155-4.12403c.0685-.54802.56829-.93675 1.11628-.86824ZM13.7422 14l.5-4.00002h-3.9844L9.75778 14h3.98442Z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
        <div
          class="group-two relative cursor-pointer"
        >
          <div
            class="z-2 pointer-events-none absolute -top-10 grid grid-cols-1 grid-rows-2 justify-items-center rounded opacity-0 group-two-hover:opacity-100"
          >
            <div
              class="hidden w-fit cursor-none rounded bg-gray-400 px-4 py-2 group-two-hover:block"
            >
              <label
                class="whitespace-nowrap font-semibold text-white"
              >
                Send to...
              </label>
            </div>
            <div>
              <svg
                fill="none"
                height="8"
                viewBox="0 0 17 8"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 0L0.5 0L7.08579 6.58579C7.86684 7.36684 9.13316 7.36684 9.91421 6.58579L16.5 0Z"
                  fill="#999999"
                />
              </svg>
            </div>
          </div>
          <button
            aria-label="Send to..."
            class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50"
          >
            <svg
              class="h-6 w-6 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M13 7.41421V14c0 .5523-.4477 1-1 1s-1-.4477-1-1V7.41421l-2.29289 2.2929c-.39053.39049-1.02369.39049-1.41422 0-.39052-.39053-.39052-1.02369 0-1.41422l4.00001-4c.3905-.39052 1.0237-.39052 1.4142 0l4 4c.3905.39053.3905 1.02369 0 1.41422-.3905.39049-1.0237.39049-1.4142 0L13 7.41421ZM6 14c0-.5523-.44772-1-1-1s-1 .4477-1 1v3c0 1.6569 1.34315 3 3 3h10c1.6569 0 3-1.3431 3-3v-3c0-.5523-.4477-1-1-1s-1 .4477-1 1v3c0 .5523-.4477 1-1 1H7c-.55228 0-1-.4477-1-1v-3Z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
        <div
          class="group-two relative cursor-pointer"
        >
          <div
            class="z-2 pointer-events-none absolute -top-10 grid grid-cols-1 grid-rows-2 justify-items-center rounded opacity-0 group-two-hover:opacity-100"
          >
            <div
              class="hidden w-fit cursor-none rounded bg-gray-400 px-4 py-2 group-two-hover:block"
            >
              <label
                class="whitespace-nowrap font-semibold text-white"
              >
                More...
              </label>
            </div>
            <div>
              <svg
                fill="none"
                height="8"
                viewBox="0 0 17 8"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 0L0.5 0L7.08579 6.58579C7.86684 7.36684 9.13316 7.36684 9.91421 6.58579L16.5 0Z"
                  fill="#999999"
                />
              </svg>
            </div>
          </div>
          <button
            aria-label="More..."
            class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50"
          >
            <svg
              class="h-6 w-6 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M14 19c0-1.1046-.8954-2-2-2s-2 .8954-2 2 .8954 2 2 2 2-.8954 2-2Zm-2-9c1.1046 0 2 .8954 2 2s-.8954 2-2 2-2-.8954-2-2 .8954-2 2-2Zm0-7c1.1046 0 2 .89543 2 2s-.8954 2-2 2-2-.89543-2-2 .8954-2 2-2Z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 group-one-hover:opacity-100"
      >
        13:00
      </div>
      <div
        class="flex w-full flex-col space-y-2 rounded py-1 pl-3 pr-2 group-one-hover:bg-gray-50"
      >
        <div
          class="leading-6"
        >
          <strong>
            A bold test message
          </strong>
          <span>
            with some more text
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
